lib/pf/cluster.pm

```diff
@@ -396,6 +408,205 @@ sub call_server {
+        
+        # Can't quorum using 2 hosts
+        if(scalar(@cluster_hosts) > 2) {
+            my $half = (scalar(@cluster_hosts) / 2);
+            my $quorum = int($half) == $half ? $half + 1 : ceil($half);
+
+            my $servers_count = 0;
+            # Figure out which servers have the quorum on the version ID
+            while(my ($version, $servers) = each(%$versions_map)) {
+                if (scalar(@$servers) > $servers_count) {
+                    $servers_count = scalar(@$servers);
+                    $quorum_version = $version;
+                }
+            }
+
+            # Ensuring they have quorum and that its not the dead servers that have quorum (through the version not being 0)
+            if ($quorum_version == 0) {
+                get_logger->warn("There are more dead servers than alive ones with the same version. Most recent configuration will be selected.");
+                goto SYNC_MOST_RECENT;
+            }
+            elsif($servers_count >= $quorum) {
+                get_logger->info("Quorum found between servers : ".join(',', @{$versions_map->{$quorum_version}}));
+                goto SYNC_QUORUM;
+            }
+            else {
+                get_logger->warn("Failed to find quorum in the cluster. Most recent configuration will be selected.");
+                goto SYNC_MOST_RECENT;
+            }
+        }
+        else {
+            get_logger->info("Quorum not possible in 2 servers clusters. Most recent configuration will be selected.");
+            goto SYNC_MOST_RECENT;
+        }
+    }
+    else {
+        get_logger->info("All servers running the same configuration version. (".join(',', keys(%$servers_map)).") have been checked.");
+        return;
+    }
+
+    # If we're here, we should return as the gotos below should be called directly
+    return;
+
+    SYNC_MOST_RECENT:
+
+    my $latest = [sort(keys(%$versions_map))]->[-1];
+
+    if($latest == $version) {
+        get_logger->info("This server is part of the nodes that are at the latest version. Synching from this node as master.");
+        sync_config_as_master();
+    }
+    else {
+        # pick the first server of the ones that are at that version and remotely call the sync on it
+        my $server = $versions_map->{$latest}->[0];
+        get_logger->info("Using $server from the servers holding running the latest version to sync as the master.");
+        call_server($server, 'sync_config_as_master');
+    }
+
+    return;
+
+    SYNC_QUORUM:
+
+    if($quorum_version == $version) {
+        get_logger->info("This server is part of the servers that have the quorum. Synching from this node as master.");
+        sync_config_as_master();
+    }
+    else {
+        my $server = $versions_map->{$quorum_version}->[0];
+        get_logger->info("Using $server from the servers holding quorum to sync as the master.");
+        call_server($server, 'sync_config_as_master');
+    }
+
+    return;
+
+}
```

Comment:
As we discussed, the gotos are there so that the logic has access to the scope of the subroutine in order to dry up the code in case additional variables would be needed